server/src/inmantals/server.py

```diff
@@ -64,9 +65,149 @@ def __init__(self, message: str) -> None:
         self.message = message
 
 
+class Folder:
+    """
+    Wrapper class around a folder (inmanta module or inmanta project). This folder can either be standalone or live in a
+    workspace alongside other folders.
+
+    The constructor should not be called explicitly. Instantiation is done through the unpack_workspaces method
+    :param folder_uri: uri of the folder that is assumed to live in a workspace
+    :param name: name of the folder
+    :param handler: reference to the InmantaLSHandler responsible for this folder
+    """
+
+    folder_uri: URI
+    name: str
+    inmanta_project_dir: Union[Optional[tempfile.TemporaryDirectory], str]
+    handler: "InmantaLSHandler"
+
+    def __init__(self, folder_uri: URI, name: str, handler: "InmantaLSHandler"):
+        self.folder_uri = folder_uri
+        self.name = name
+        self.handler = handler  # Keep a reference to the handler for cleanup
+
+        # Check that we are working inside an existing project:
+        project_file: str = os.path.join(folder_uri.path, module.Project.PROJECT_FILE)
+        if os.path.exists(project_file):
+            self.inmanta_project_dir = os.path.dirname(project_file)
+        else:
+            self.inmanta_project_dir = None
+
+    def get_project_dir(self) -> Optional[str]:
+        """
+        If this folder holds an inmanta project, this returns the path to this project.
+        If this folder holds an inmanta module, this returns the path of the temporary
+        inmanta project used to load this module.
+        """
+        if not self.inmanta_project_dir:
+            return None
+        if isinstance(self.inmanta_project_dir, str):
+            return self.inmanta_project_dir
+        else:
+            return self.inmanta_project_dir.name
+
+    def get_setting(self, key: str, default: Optional[str] = None) -> Optional[str]:
+        """
+        Get an inmanta extension setting for this folder
+        """
+        if key == "compiler_venv_path":  # support for versions of core that require a compiler venv.
+            return os.path.join(self.get_folder_path(), ".env-ls-compiler")
+        try:
+            # Look for some custom settings for this folder
+            with open(os.path.join(self.get_folder_path(), ".vscode", "settings.json")) as settings_file:
+                settings: dict[str, str] = json.load(settings_file)
+                return settings.get(key, default)
+        except FileNotFoundError:
+            return default
+
+    @classmethod
+    def unpack_workspaces(cls, workspace_folders: Sequence[object], ls_handler: "InmantaLSHandler") -> Dict[str, "Folder"]:
+        return {folder["uri"]: Folder(URI(folder["uri"]), folder["name"], ls_handler) for folder in workspace_folders}
+
+    def cleanup(self) -> None:
+        logger.info(f"calling cleanup for {self}")
```

Comment:
Generally we shouldn't use f-strings for logging, recall Wouter's lightning talk at the last team week. There are some other logging instances where this should be applied, I didn't mark them all.
```suggestion
        logger.info("calling cleanup for %s", self)
```

Reply:
I'll change all instances. I remember him giving a talk about this, but I don't recall the exact issue with this. I looked around a bit and found that for performance reasons the %s style is better. Was there another reason not to use f-strings ?

Reply:
In general f-strings are fine, but the logging framework uses `%s` intrinsically (see how my suggestion has a comma rather than a `%`, so we really pass the values to the logging framework). There are some performance considerations, but more importantly it's just a good practice to do it as the logging framework intended. The worst scenario is if you would use a mixed style: `logger.info(f"hello {world} %s", "!")`. If `world` happens to contain another `%s`, this will result in an exception. So it's way safer than in some low level languages, but it's still something to take into account.